gui, lib/versioner: Allow to configure staggered versioning intervals (fixes #8352)

tomasz1986 opened the pull request:
gui, lib/versioner: Allow to configure staggered versioning intervals (fixes #8352)

Currently, there are four intervals used in the staggered versioning,
with all of them but the Maximum Age being hard-coded. With this change,
the user is able to configure all intervals to their liking. Due to the
complexity of the configuration, this option is intended to be used by
advanced users only.

In addition to the above, also:

1. Add yet another interval for more fine-grained configuration
   possibilities. By default, the new interval activates after one year,
   keeping one version per month. However, because the default Maximum
   Age is also set to one year, this fifth interval is actually not
   utilised unless the user increases the Maximum Age to a higher value.

2. Fix a small bug in the Staggered Versioning test where the default
   time period for the third interval was supposed to be set to 30 days,
   but in reality it was set to just 7 days.

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

### Screenshots

![image](https://github.com/syncthing/syncthing/assets/5626656/8e38238d-a268-4822-92e5-9bd5bc5ec122)
![image](https://github.com/syncthing/syncthing/assets/5626656/7740c58a-2c64-41d7-bd5f-f4cec4a51f4e)

tomasz1986 committed:
gui, lib/versioner: Allow to configure staggered versioning intervals (fixes #8352)

Currently, there are four intervals used in the staggered versioning,
with all of them but the Maximum Age being hard-coded. With this change,
the user is able to configure all intervals to their liking. Due to the
complexity of the configuration, this option is intended to be used by
advanced users only.

In addition to the above, also:

1. Add yet another interval for more fine-grained configuration
   possibilities. By default, the new interval activates after one year,
   keeping one version per month. However, because the default Maximum
   Age is also set to one year, this fifth interval is actually not
   utilised unless the user increases the Maximum Age to a higher value.

2. Fix a small bug in the Staggered Versioning test where the default
   time period for the third interval was supposed to be set to 30 days,
   but in reality it was set to just 7 days.

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 commented:
I will be grateful for any tips on how to simplify this code 😅.

tomasz1986 commented on gui/default/syncthing/folder/editFolderModalView.html:
```diff
@@ -115,19 +115,89 @@
           </div>
-          <div class="form-group" ng-if="currentFolder._guiVersioning.selector=='staggered'" ng-class="{'has-error': folderEditor.staggeredMaxAge.$invalid && folderEditor.staggeredMaxAge.$dirty}">
-            <p class="help-block"><span translate>Files are moved to date stamped versions in a .stversions directory when replaced or deleted by Syncthing.</span> <span translate>Versions are automatically deleted if they are older than the maximum age or exceed the number of files allowed in an interval.</span></p>
-            <p translate class="help-block">The following intervals are used: for the first hour a version is kept every 30 seconds, for the first day a version is kept every hour, for the first 30 days a version is kept every day, until the maximum age a version is kept every week.</p>
```
I had to break this very long translation, so unfortunately parts of it will have to be translated again, but this time I've also tried to split this into separate strings, so that it should be easier to deal with from now on if there are any small changes required.

tomasz1986 committed:
maxAge name fix

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
remove unnecessary brackets

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

calmh commented:
Didn't look at any code yet, but I can say that exposing config values in the range of millions of seconds in the non-advanced GUI certainly doesn't seem ideal.

tomasz1986 commented:
Yeah, I know about that. I originally just tried using time scales that are mentioned in the code comments (day/hour/week), however that didn't allow for such fine-grained control as the current iteration (which the user in the enhancement request wanted).

I think there could be two solutions to this:

- Leave the first interval at seconds, change the second one to hours, and then the rest to days. The Maximum Age, i.e. the last period, is already set using days, so I don't think it makes sense to use any larger time units than days.

or

- make it so that if you click on the time unit label, it cycles through seconds/hours/days/weeks/months/years, etc. which would allow to set the larger intervals much easier while still allowing for the current fine-grained control.

The second approach would probably be ideal and much more flexible, but I'm not sure if I can pull this off in a reasonable amount of time without extreme code complexity, so for now I'll probably just try out the first one and see how it works.

tomasz1986 committed:
Merge branch 'main' into tomasz86/gui/improve-staggered-versioning-#8352

tomasz1986 committed:
wip

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
wip

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
wip

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
wip

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
wip

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
wip

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
wip

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
wip

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
wip

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
wip

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

tomasz1986 committed:
Merge branch 'main' into tomasz86/gui/improve-staggered-versioning-#8352

uok commented:
The information about default intervals is helpful! 
My suggestion is to make the interval configuration available for advanced users by clicking a link "configure intervals".
Maybe keep it in the style of a sentence, e.g.: "for the first ___ minutes keep a version for every ___ seconds" (___ = input box)
In the folder overview the icon/tooltips are only needed if settings differ from default values.

tomasz1986 commented:
> The information about default intervals is helpful!

The information is already there right now! It's just because I added one more interval, I had to modify the sentence, and I though that the whole thing would be easier to read and understand when split into a bullet point list.

> My suggestion is to make the interval configuration available for advanced users by clicking a link "configure intervals".
> Maybe make input it in the style of a sentence, e.g.: "for the first ___ minutes keep a version for every ___ seconds" (___ = input box)

Do you think it would be better to have it displayed in a separate modal? It's definitely possible to do, but this part is only visible when you're using Staggered Versioning already, so that's why I just added it to the main File Versioning configuration window together with the other, already existing options. Of course, it's not visible when using other types of versioning (or no versioning at all).

> In the folder overview the icon/tooltips are only needed if settings differ from default values.

This has recently been changed in https://github.com/syncthing/syncthing/pull/8987. We now show all values regardless of whether they are default or not.

calmh commented:
Maybe we could have a time picker of some kind.

![Unknown](https://github.com/syncthing/syncthing/assets/125426/3b0e0fe2-05a9-4257-94dd-0f84967cb883)
